File: leetcode/tester.py
from pathlib import Path
from time import perf_counter

class Tester:
    def __init__(self, solution):
        self.name = next(filter(lambda it: not it.startswith("_"), dir(solution)))
        self.fun = getattr(solution, self.name)
        self.total = 0
        self.fail = 0
        self.time = 0

    def test(self, expected, *args, **kwargs):
        start = perf_counter()
        output = self.fun(*args, **kwargs)
        self.time += perf_counter() - start

        isCorrect = output == expected
        self.total += 1

        print(f"Test {self.total} ... {isCorrect}")
        if not isCorrect:
            print("Input:", args, kwargs)
            print("Output:  ", output)
            print("Expected:", expected)
            self.fail += 1
        print()

    def report(self):
        print(f"{self.name} passed {self.total - self.fail} / {self.total} tests")
        print(f"Ran in {round(self.time*1000, 4)}ms")
        print()


class ObjectTester:
    def __init__(self, filepath):
        self.module = __import__(Path(filepath).stem)
        self.total = 0
        self.fail = 0
        self.time = 0
    
    def test(self, expected, ops, args):
        start = perf_counter()
        obj = getattr(self.module, ops[0])(*args[0])
        outputs = [None]
        outputs.extend(getattr(obj, op)(*arg) for op, arg in zip(ops[1:], args[1:]))
        self.time += perf_counter() - start

        for op, arg, output, expect in zip(ops, args, outputs, expected):
            isCorrect = output == expect
            self.total += 1

            print(f"Test {self.total} ... {isCorrect}")
            if not isCorrect:
                print("Input:", op, arg)
                print("Output:  ", output)
                print("Expected:", expect)
                self.fail += 1
            print()

    def report(self):
        print(f"Passed {self.total - self.fail} / {self.total} tests")
        print(f"Ran in {round(self.time*1000, 4)}ms")
        print()
            

File: leetcode/tester.js
export default class Tester {
  constructor(fun) {
    this.fun = fun
    this.total = 0
    this.fail = 0
    this.time = 0
  }

  get pass() {
    return this.total - this.fail
  }

  test(expected, ...input) {
    const start = performance.now()
    const output = this.fun(...input)
    this.time += performance.now() - start

    const isCorrect = output === expected

    console.log('Test', this.total, '...', isCorrect)

    if (!isCorrect) {
      console.error('Output:', output)
      console.error('Expected:', expected)
      ++this.fail
    }

    console.log('')
    ++this.total
  }

  report() {
    console.log(`Passed ${this.pass} / ${this.total} tests
`)
    console.log(`Ran in ${this.time.toFixed(4)}ms`)
  }
}

const prettyPrint = (arr) => {
  const n = arr.length
  process.stdout.write(`(${n}) [`)
  for (const [i, it] of arr.entries()) {
    process.stdout.write(it < 10 ? ' ' : '')
    process.stdout.write(it.toString())
    process.stdout.write(i < n - 1 ? ' ' : '')
  }
  process.stdout.write(']\n')
}

const arange = (start, end) =>
  Array(end - start)
    .fill()
    .map((_, idx) => idx + start)

const swap = (arr, i, j) => {
  ;[arr[i], arr[j]] = [arr[j], arr[i]]
}

export { prettyPrint, arange, swap }
